File: coco/src/main/java/coco/ep/m6/Add.java
package coco.ep.m6;

public interface Add<FT> extends coco.ep.m5.Add<FT>, Exp<FT> {

    default Boolean equals(coco.ep.Exp<FT> other) {
        return convert(this).astree().equals(convert(other).astree());
    }

    default Boolean isAdd(coco.ep.Exp<FT> left, coco.ep.Exp<FT> right) {
        return convert(left).eql(getLeft()) && convert(right).eql(getRight());
    }

    default Boolean eql(coco.ep.Exp<FT> that) {
        return convert(that).isAdd(convert(getLeft()), convert(getRight()));
    }
}

File: coco_csharp/CoCo/EP/M6/Add.cs
namespace EP.M6 {
    public interface Add<FT> : Exp<FT>, EP.M5.Add<FT> {
        bool Exp<FT>.equals(EP.Exp<FT> other) {
            return convert(this).astree().Equals(convert(other).astree());
        }

        bool Exp<FT>.isAdd(EP.Exp<FT> left, EP.Exp<FT> right) {
            return convert(left).eql(getLeft()) && convert(right).eql(getRight());
        }

        bool Exp<FT>.eql(EP.Exp<FT> that) {
            return convert(that).isAdd(convert(getLeft()), convert(getRight()));
        }
    }
}